numpy/core/src/multiarray/item_selection.c

```diff
@@ -2128,6 +2129,82 @@ count_nonzero_bytes_384(const npy_uint64 * w)
 }
 
+#if NPY_SIMD
+
+/* Count the zero bytes between `*d` and `end`, updating `*d` to point to where to keep counting from. */
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u8
+count_zero_bytes_u8(const npy_uint8 **d, const npy_uint8 *end, npy_uint8 max_count)
+{
+    const npyv_u8 vone = npyv_setall_u8(1);
+    const npyv_u8 vzero = npyv_zero_u8();
+
+    npy_intp lane_max = 0;
+    npyv_u8 vsum8 = npyv_zero_u8();
+    while (*d < end && lane_max <= max_count - 1) {
+        npyv_u8 vt = npyv_cvt_u8_b8(npyv_cmpeq_u8(npyv_load_u8(*d), vzero));
+        vt = npyv_and_u8(vt, vone);
+        vsum8 = npyv_add_u8(vsum8, vt);
+        *d += npyv_nlanes_u8;
+        lane_max += 1;
+    }
+    return vsum8;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u16
+count_zero_bytes_u16(const npy_uint8 **d, const npy_uint8 *end, npy_uint16 max_count)
+{
+    npyv_u16 vsum16 = npyv_zero_u16();
+    npy_intp lane_max = 0;
+    while (*d < end && lane_max <= max_count - 2*NPY_MAX_UINT8) {
+        npyv_u8 vsum8 = count_zero_bytes_u8(d, end, NPY_MAX_UINT8);
+        npyv_u16x2 part = npyv_expand_u16_u8(vsum8);
+        vsum16 = npyv_add_u16(vsum16, npyv_add_u16(part.val[0], part.val[1]));
+        lane_max += 2*NPY_MAX_UINT8;
+    }
+    return vsum16;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u32
+count_zero_bytes_u32(const npy_uint8 **d, const npy_uint8 *end, npy_uint32 max_count)
+{
+    npyv_u32 vsum32 = npyv_zero_u32();
+    npy_intp lane_max = 0;
+    while (*d < end && lane_max <= max_count - 2*NPY_MAX_UINT16) {
+        npyv_u16 vsum16 = count_zero_bytes_u16(d, end, NPY_MAX_UINT16);
+        npyv_u32x2 part = npyv_expand_u32_u16(vsum16);
+        vsum32 = npyv_add_u32(vsum32, npyv_add_u32(part.val[0], part.val[1]));
+        lane_max += 2*NPY_MAX_UINT16;
+    }
+    return vsum32;
+}
+/*
+ * Counts the number of non-zero values in a raw array.
+ * The one loop process is as follows(take SSE2 with 128bits vector for example):
+ *          |------------16 lanes---------|          
+ * [vsum8]  255 255 255 ... 255 255 255 255 count_zero_bytes_u8: counting 255*16 elements
+ *                          !!
+ *           |------------8 lanes---------|          
+ * [vsum16]  65535 65535 65535 ...   65535  count_zero_bytes_u16: counting 65535*8 elements
+ *                          !!
+ *           |------------4 lanes---------|          
+ * [vsum32]  1073741824   ...   1073741824  count_zero_bytes_u32(overflow control): counting 2**32-1 elements
+ *                          !!
+ *                        2**32-1           count_zero_bytes
+*/
+static NPY_INLINE NPY_GCC_OPT_3 npy_intp
+count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
+{
+    npy_intp zero_count = 0;
+    const npy_uint8 *end = d + unrollx;
+    while (d < end) {
+        // The npyv_nlanes_u32 factor ensures that the sum of all lanes still fits in a uint32
+        npyv_u32 vsum32 = count_zero_bytes_u32(&d, end, NPY_MAX_UINT32 / npyv_nlanes_u32);
+        zero_count += npyv_sum_u32(vsum32);
```

Comment:
```suggestion
    npyv_u16x2 vsum16 = count_zero_bytes_u16(d, end, NPY_MAX_UINT16);
    npyv_u32x2 sum_32_0 = npyv_expand_u32_u16(vsum16.val[0]);
    npyv_u32x2 sum_32_1 = npyv_expand_u32_u16(vsum16.val[1]);
    zero_count += npyv_sum_u32(sum_32_0.val[0]);
    zero_count += npyv_sum_u32(sum_32_0.val[1]);
    zero_count += npyv_sum_u32(sum_32_1.val[0]);
    zero_count += npyv_sum_u32(sum_32_1.val[1]);
```

Reply:
![image](https://user-images.githubusercontent.com/15245051/102732252-2a9b7d80-4375-11eb-915c-c38aa0441f98.png)
Please correct me if i am wrong, the new solution only handles `(2**16-1)*16` elements in one loop, while the previous one handles `2**32-1` elements, the iterations `x2` is good, but I think that `count_zero_bytes_u32 `should not be removed from the perspective of efficiency.

Reply:
> but I think that count_zero_bytes_u32 should not be removed from the perspective of efficiency.

wouldn't affect performance, the idea behind level u16 is to reduce summing the vector.
if you want to improve the performance more, try to unroll u8 level by x4.

>  the npyv_expand_u32_u16 operation is used just to prevent summation overflow

sorry I wrote the suggestion in a hurry, I edited it only one `npyv_sum_u32()` is used.

Reply:
> but in previous solution expansion is used to count more elements, which is the key task here.

Which one do you refer to? I'm getting confused.

Reply:
I prefer to reserve `count_zero_bytes_u32` as shown below:
![image](https://user-images.githubusercontent.com/15245051/102762388-2beb9b00-43b3-11eb-8618-09ab71499cd7.png)

Reply:
I think there's a misunderstanding here, what I can understand from the code:

  - `count_nonzero_bytes_u8()`  to reduce calling `npyv_expand_u16_u8()` and  `npyv_sum_u32()`
  - `count_nonzero_bytes_u16()` to reduce calling `npyv_expand_u32_u16()`, `npyv_sum_u32()`
  - `count_nonzero_bytes_u32()` another level to reduce calling `npyv_sum_u32()` not because "needs to avoid overflow"

here an example for the simd loop without any block-level so you can get what I understand from your code:

```C
static NPY_INLINE NPY_GCC_OPT_3 npy_intp
count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
{
    npy_intp zero_count = 0;
    for (; unrollx > 0; unrollx -= npyv_nlanes_u8, d += npyv_nlanes_u8) {
        npyv_u8 cmp = npyv_cvt_u8_b8(npyv_cmpeq_u8(npyv_load_u8(d), npyv_zero_u8()));
        npyv_u8 one = npyv_and_u8(cmp, npyv_setall_u8(1));

        npyv_u16x2 one_u16 = npyv_expand_u16_u8(vsum8);
        npyv_u32x2 one_u32 = npyv_expand_u32_u16(npyv_add_u16(one_u16.val[0], one_u16.val[1]));
        zero_count += npyv_sum_u32(npyv_add_u32(one_u32.val[0], one_u32.val[1]));
    }
    return zero_count;
}
````

Now, Am I missing something?

Reply:
The non-block-level code is fine, basically there has four loops here:
1. `count_zero_bytes_u8`: get the maximum count that u8 type can hold.
2. `count_zero_bytes_u16`: get the maximum count that u16 type can hold.
3. `count_zero_bytes_u32`: get the maximum count that u32 type can hold.
4. `count_zero_bytes`: get the maximum count that u64 type(AKA `npy_uintp`) can hold.

Reply:
@Qiyu8, the step 3 you describe there isn't quite the (correct) code you've incremented - the `u32` version fills a vector with values whose _sum_ does not exceed a u32

Reply:
In the previous example, I was trying to explain the motives for creating nested loops.
The original OpenCV code was trying to reduce the calls of expensive intrinsics
as much as possible and increase the cheaper ones.

cheap intrinsics:
 - using compare equal since almost all archs don't have native support "not equal".
 - integer addition.
 - bitwise

Most archs can execute multiple instructions for the
the above operations per one clock cycle but on the other hand
expand and reduce sum(the worst) takes more latency and throughput.

The negative side of nested loops is the "integer overflow" which leads
to putting an iterations limit to the inner loops but wait loops involve jmps
and jmps may lead to flushing the pipeline  so you
should be aware that the inner loop should save more cycles than what the flushing can spend.

> what I don't fully understand is that your suggested code removes step 3.

because I think there's no performance gain from it, `count_zero_bytes_u16` already reduces enough calls of "reduce sum".

Reply:
The performance looks fine after remove inner loop.
```
       before           after         ratio
     [d7a75e8e]       [c5daaf06]
     <master>         <countnz>
-      6.47±0.1ms       5.48±0.2ms     0.85  bench_core.CountNonzero.time_count_nonzero(2, 1000000, <class 'int'>)
-     3.27±0.08ms      2.69±0.08ms     0.82  bench_core.CountNonzero.time_count_nonzero(1, 1000000, <class 'int'>)
-         101±3μs         82.6±2μs     0.82  bench_core.CountNonzero.time_count_nonzero(3, 10000, <class 'int'>)
-      10.1±0.3ms       8.07±0.1ms     0.80  bench_core.CountNonzero.time_count_nonzero(3, 1000000, <class 'int'>)
-         127±3μs         92.8±4μs     0.73  bench_core.CountNonzero.time_count_nonzero(2, 1000000, <class 'bool'>)
-        215±10μs          142±3μs     0.66  bench_core.CountNonzero.time_count_nonzero(3, 1000000, <class 'bool'>)
-        64.6±1μs         41.7±2μs     0.65  bench_core.CountNonzero.time_count_nonzero(1, 1000000, <class 'bool'>)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```